### PR TITLE
[WIP] Better event handling for Instance related tasks

### PR DESCRIPTION
**Pull request by Xarthisius**
~With this PR status of an *Instance* is properly updated according to status of all related tasks: `create_volume`, `launch_container`, `shutdown_container`, `remove_volume`. In particular:~

1. ~When the first task (`create_volume`) fails, instance status is updated to `ERROR`.~
2. ~When `DELETE /instance` is called, *Instance* status is set to `DELETING` whenever `shutdown_instance` and `remove_volume` are scheduled or running~
3. ~*Instance* is only removed when `remove_volume` finishes with a success.~

~Additionally, in order to indicate that `DELETE /instance` is an asynchronous task, the endpoint now returns 202 ACCEPTED code.~

With this PR status of an *Instance* is updated according to status of most related tasks:
`create_volume`, `launch_container`, `shutdown_container`. Currently all that's being done for `shutdown_container` is setting *Instance* status to `DELETING`. Instances in `DELETING` and `ERROR` state are not taken into account for calculating instance cap limit.

As [suggested](https://github.com/whole-tale/dashboard/issues/213#issuecomment-447056036) by @bodom0015, introduction of DELETING state will allow us to workaround https://github.com/whole-tale/dashboard/issues/213.

### How to test?
1. Modify `gwvolman`'s `create_volume` so that it always fails (e.g. raise any error)
2. Launch a Tale
3. Confirm that newly created instance ends up with status `ERROR` (2)
4. Revert 1.
5. Add `time.sleep(60)` to either `shutdown_instance` or `remove_volume` in `gwvolman`
6. Launch a Tale
7. Delete an Instance after it's started.
8. Confirm that instance retains `DELETING` (3) status as long as `shutdown_instance` or `remove_volume` are running.

### CODE REVIEW

File: plugin_tests/instance_test.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 import time
 import json
-import copy
```
Comment (bodom0015) [inline plugin_tests/instance_test.py]: Thank you for cleaning this up! My bad, debugging oversight :X

File: server/rest/instance.py [hunk 2]
```diff
@@ -156,12 +157,18 @@ def updateInstance(self, instance):
...
+               default=False, required=False, dataType='boolean')
         .errorResponse('ID was invalid.')
         .errorResponse('Write access was denied for the instance.', 403)
     )
-    def deleteInstance(self, instance, params):
-        self.model('instance', 'wholetale').deleteInstance(
-            instance, self.getCurrentToken())
+    def deleteInstance(self, instance, force):
+        if force:
+            self.model('instance', 'wholetale').remove(instance)
```
Comment (bodom0015) [inline server/rest/instance.py]: I may be overthinking things here, but don't we still want to try to call the handler to cleanup the underlying container when forcing? Unless `self.model('instance', 'wholetale').remove(instance)` would somehow trigger an event to clean things up or something..

The description is explicit about what happens to the Instance resource, but it is not clear about what happens to the container itself.

**Comment by craig-willis** [comment]
Tested again with latest changes and it appears to be working.  As discussed, the `sleep(60)` doesn't work because of the `TASK_TIMEOUT=15` in `deleteInstance().` After setting to `sleep(10)` it worked as expected.

Noting two other issues (@bodom0015):
* When delete fails, Dashboard reports nothing.
* Even when the instance status is set to DELETING, the UI still shows the instance. 
* Even after the instance is removed, the instance still displays in UI

**Comment by Xarthisius** [comment]
Stale and rotten...

**Pull request closed without merging**